sdks/java/io/cdap/src/main/java/org.apache.beam.sdk.io.cdap/Plugin.java

```diff
@@ -0,0 +1,176 @@
+    /**
+     * Sets InputFormat or OutputFormat class for a plugin.
+     */
+    public void setFormatClass(Class<?> formatClass) {
+        this.formatClass = formatClass;
+    }
+
+    /**
+     * Gets InputFormat or OutputFormat class for a plugin.
+     */
+    public Class<?> getFormatClass() {
+        return formatClass;
+    }
+
+    /**
+     * Sets InputFormatProvider or OutputFormatProvider class for a plugin.
+     */
+    public void setFormatProviderClass(Class<?> formatProviderClass) {
+        this.formatProviderClass = formatProviderClass;
+    }
+
+    /**
+     * Gets InputFormatProvider or OutputFormatProvider class for a plugin.
+     */
+    public Class<?> getFormatProviderClass() {
+        return formatProviderClass;
+    }
+
+    /**
+     * Sets a plugin config.
+     */
+    public Plugin withConfig(PluginConfig pluginConfig) {
+        this.pluginConfig = pluginConfig;
+        return this;
+    }
+
+    /**
+     * Gets a plugin config.
+     */
+    public PluginConfig getPluginConfig() {
+        return pluginConfig;
+    }
+
+    /**
+     * Sets a plugin Hadoop configuration.
+     */
+    public Plugin withHadoopConfiguration(Class<?> OutputFormatKeyClass,
+                                              Class<?> OutputFormatValueClass) {
+        PluginConstants.PluginType pluginType = getPluginType();
+
+        PluginConstants.Format formatType =
+                pluginType == PluginConstants.PluginType.SOURCE
+                ? PluginConstants.Format.INPUT
+                : PluginConstants.Format.OUTPUT;
+
+        PluginConstants.Hadoop hadoopType =
+                pluginType == PluginConstants.PluginType.SOURCE
+                ? PluginConstants.Hadoop.SOURCE
+                : PluginConstants.Hadoop.SINK;
+
+        this.hadoopConfiguration = new Configuration(false);
+
+        this.hadoopConfiguration.setClass(hadoopType.getFormatClass(), formatClass, formatType.getFormatClass());
+        this.hadoopConfiguration.setClass(hadoopType.getKeyClass(), OutputFormatKeyClass, Object.class);
+        this.hadoopConfiguration.setClass(hadoopType.getValueClass(), OutputFormatValueClass, Object.class);
+
+        return this;
+    }
+
+    /**
+     * Sets a plugin Hadoop configuration.
+     */
+    public Plugin withHadoopConfiguration(Configuration hadoopConfiguration) {
+        this.hadoopConfiguration = hadoopConfiguration;
+
+        return this;
+    }
+
+    /**
+     * Gets a plugin Hadoop configuration.
+     */
+    public Configuration getHadoopConfiguration() {
+        return hadoopConfiguration;
+    }
+
+    /**
+     * Sets a plugin type.
+     */
+    public void setPluginType(PluginConstants.PluginType pluginType) {
+        this.pluginType = pluginType;
+    }
+
+    /**
+     * Gets a plugin type.
+     */
+    public PluginConstants.PluginType getPluginType() {
+        return pluginType;
+    }
+
+    /**
+     * Validates plugin fields.
+     */
+    public void validatePluginClass() {
```

Comment:
Maybe it will be better to implement validating in `PluginBuilder`, and do it before creating `Plugin` instance ?